### PR TITLE
Proof of concept: use Mithril's mocking functions to query actual DOM

### PR DESCRIPTION
**Pull request by tombh**
Great library! Thanks.

I've been playing around to see if you can use this against the DOM as generated by `m.route()`. The mithril.js repo itself comes with great mocks for `window` and `window.XMLHttpRequest`, see: https://github.com/lhorie/mithril.js/blob/next/tests/mock.js So it's already easy to render mithril modules into a fake DOM.

So taking it a step further I've added an extra cssauron language for a rendered DOM. I've also added some tests that show how you can now also test modules by visiting URLs (eg; `m.route('/')`) and it even works with URL params, eg; `/route/:param`. Mithril's mocking functions also stub out AJAX requests, so I've also added a test that shows how to render a view with a fake AJAX response. Who needs PhantomJS!

There's definitely a lot more tests and tidying up that could be done to this, but I'm just PR'ing it now to see if this is something you'd be interested in? If so, I wonder if @lhorie would be up for packaging `mock.js` along with the npm package? So we didn't have to copy and paste it into this repo every time it updates.

### CODE REVIEW

**Comment by StephanHoyer** [comment]
This is great stuff.

I tried something like this when introducing components but didn't quite get it working. I tried to use jsdom/simple-dom for this. But you solution is much cleaner and probably more robust.

It also uses mithrils rendering for generating stuff. So changes in mithril (e. G. intro of custom attributes) would not require this lib to reimplement this.

**Comment by StephanHoyer** [comment]
If you get all existing tests working, I might consider switching the whole thing to your solution.

**Comment by tombh** [comment]
Thanks :)

Okay, I'll see if I can get the full the test suite working with this method.

So, say if you do move over to this method, how will you provide backwards compatibility? Because any projects currently running their test suites with this library will suddenly have failing tests if they upgrade.

**Comment by StephanHoyer** [comment]
Why that? I think one could do this without changing the API I thought

You must not use `m.route` to start mithril to work. you can `m.mount` components right away.

**Comment by tombh** [comment]
I agree that you can do this without changing the API. But it would mean supporting 2 cssauron languages (as I currently do in this PR) and possibly 2 test suites.

`mq()` currently expects to be given a view, module or a pre-built mithril virtual DOM tree, which uses custom object properties such as `children` and `parent`. However, when you call `m.render`, `m.mount` or `m.route` on a virtual DOM object, it gets converted into a native DOM object, which is attached to `window.document`. This DOM tree has a different syntax (the one used by browsers), where it has properties such as `childNodes` and `parentNode`.

If we moved over to just using the method proposed in this PR then all existing `mq()` calls would fail because it would no longer support virtual DOM trees.

Here are examples of the 2 DOM trees.
**Mithril virtual DOM**
```JS
{ rootEl: 
   { tag: 'div',
     attrs: { className: 'root' },
     children: 
      [ [Object],
        [Object],
        [Object],
        'Inner String',
        [Object],
        [Object],
        [Object],
        [Object],
        undefined,
        10,
        [Object],
        [Object],
        [Object] ] }
```
**Rendered DOM**
```JS
[ { style: {},
    childNodes: [ [Object] ],
    nodeType: 1,
    nodeName: 'LI',
    appendChild: [Function],
    removeChild: [Function],
    replaceChild: [Function],
    insertBefore: [Function],
    insertAdjacentHTML: [Function],
    setAttribute: [Function],
    setAttributeNS: [Function],
    getAttribute: [Function],
    addEventListener: [Function],
    removeEventListener: [Function],
    parentNode: 
     { style: {},
       childNodes: [Object],
       nodeType: 1,
       nodeName: 'UL',
       appendChild: [Function],
       removeChild: [Function],
       replaceChild: [Function],
       insertBefore: [Function],
       insertAdjacentHTML: [Function],
       setAttribute: [Function],
       setAttributeNS: [Function],
       getAttribute: [Function],
       addEventListener: [Function],
       removeEventListener: [Function],
       parentNode: [Object] } } ]
```

Does that make sense?

**Comment by tombh** [comment]
Wait! I just thought!! We could just run `m.mount` in mintril-query's `init()` function whenever it detects receiving a virtual DOM! Of course, so simple. Is that what you meant?

That way *whatever* `mq()` is passed, internally mithril-query will be working with a mithril-mock's native DOM syntax. So we get backwards compatibility *and* the ability to pass a `document.body` to `mq()`.

:grinning:

**Comment by tombh** [comment]
Hmm, my only concern with that is that there would be no guarantee that all existing test cases out in the would work. Maybe there are subtle edge cases where querying or triggering events on a virtual DOM are different from the native DOM. I think we'd need to write some extra tests just to try and make extra sure that true backwards compatibility was being maintained.

**Comment by StephanHoyer** [comment]
yep that's what i thought... if we get rendered VDOM we can just wrap it in a function.

And yes, some extra test-cases would be good. Maybe we also can first support both ways and deprecate the old one. But I'm confident that we can get this done without breaking BC.

I have a pretty large application with good coverage, if the change don't break any tests, we can be pretty sure this works ;)

**Comment by tombh** [comment]
Ha, straight away on [the first expectation](https://github.com/StephanHoyer/mithril-query/blob/master/test.js#L35) there are 2 problems.

1) Consider this: `expect(out.first('span')).to.eql(tagEl);` Using the new method `out.first('span')` is going to have native DOM syntax and `tagEl` is going to have VDOM syntax because it is defined earlier as `tagEl = m('span');`. So even if they theoretically match (as in both DOMs represent the same element), they won't match when programatically compared. I don't see any solution to this other than continuing to maintain compatibility for both DOM syntaxes and perhaps slowly fading out the VDOM syntax over time. We could do it so that at first you have to opt-in to using native DOM syntax and then over time we make it the default whilst having to opt-in to the VDOM syntax and after some time deprecate the VDOM syntax completely.

2) The [expect.js](https://github.com/Automattic/expect.js/issues/100) project is dead. I've found a bug where it errors when comparing DOM objects in node. There is no hope of having my fix included. A quick solution for this project would be to just include the [npm xmlserialiser package](https://www.npmjs.com/package/xmlserializer). However I think a better solution would be to use [a more maintained expect library](https://github.com/mjackson/expect), or to use Chai.

**Comment by StephanHoyer** [comment]
Since the vdom objects are pretty easy one could simply write a converter
dom->vdom. I did excactly that when using Jsdom approach. Probably I
overwrote this by forcepushing. Maybe it still exists in a branch or tag.
You can definitely pull this from npm.


I'm on mobile right now. I can look in 8 h or so. Maybe you find something
by yourself.

Thanks again for the time you put in this.


Am Samstag, 20. Juni 2015 schrieb Thomas Buckley-Houston :

**Comment by tombh** [comment]
Thanks. And thank you for writing this great little project :)

Ah yes, I found it https://github.com/StephanHoyer/mithril-query/blob/3b29f75f70a89fa66ff5ab229efe8137ec11f884/index.js#L61 That looks straightforward.

```js
function toVdomEl(node) {
  var attrMap = {
    classname: 'className'
  };
  if (node.nodeName === '#text') {
    return node.nodeValue;
  }
  var vdomEl = {
    tag: node.tagName.toLowerCase()
  };
  vdomEl.attrs = node.attributes.reduce(function(attrs, attribute) {
    attrs[attrMap[attribute.name] || attribute.name] = attribute.value;
    return attrs;
  }, {});
  Object.keys(node).reduce(function(vdomEl, key) {
    if (isFunction(node[key])) {
      vdomEl.attrs[key] = node[key];
    }
    return vdomEl;
  }, vdomEl);
  vdomEl.children = getChildNodes(node).map(function(node) {
    var child = toVdomEl(node);
    child.parent = vdomEl;
    return child;
  });
  return vdomEl;
}
```

I'm a bit confused, so just to be clear: if you pass a real DOM tree syntax, like `mq(document.body)`, then the result of `mq(document.body).first('span)` will *also* be a real DOM syntax tree. Then if you pass a vDOM syntax tree, like `mq(m('span', 'sample element text'))` then internally `mq()` will use `m.render()` to generate a real DOM and then afterwards convert the result back to a vDOM, so that the result of `mq(m('span', 'sample element text')).first('span')` will be a *vDOM* syntax tree?

Why did you give up on the `simple-dom` approach then?

Also what do you think about moving away from `expect.js`? If we don't move away then we can just include the `xmlserialiser` package, but considering that the `expect.js` is dead, then I think it'd be a good opportunity to use something else.

**Comment by StephanHoyer** [comment]
I used `expect.js` ever since. Didn't know that it's dead. `expect` looks really nice. Maybe we should first change the assertion lib before continuing this.

I stopped using `simple-dom` because it was very unstable. Also it lacks of some features mithril needs ([insertAdjacentHTML](https://developer.mozilla.org/de/docs/Web/API/Element/insertAdjacentHTML)). I shimmed that, but is was q

**Comment by StephanHoyer** [comment]
Maybe we can have a chat on this on [Gitter](https://gitter.im/StephanHoyer/mithril-query?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

File: index.js [hunk 1]
```diff
@@ -33,38 +37,105 @@ function isTrusted(thing) {
...
+      }).join(' ');
+    } else {
+      return '';
     }
-    return isArray(node.children) ? node.children.filter(isString).concat('') : '';
   },
   id: function(node) {
-    if (node.attrs) {
-      return node.attrs.id;
+    if (node.id) {
```
Comment (StephanHoyer) [inline index.js]: is this if required?

File: index.js [hunk 2]
```diff
@@ -274,35 +388,32 @@ function scan(render) {
 }
 
 function init(viewOrModuleOrRootEl, scope, b, c, d, e, f, noWay) {
```
Comment (StephanHoyer) [inline index.js]: maybe we can save `arguments` here and don't use this `a, b, c`-pattern

File: index.js [hunk 3]
```diff
@@ -98,6 +189,8 @@ function scan(render) {
       if (!el) {
         return foundEls;
       }
+
+      // TODO: Is this still needed?
       if (isModule(el)) {
```
Comment (StephanHoyer) [inline index.js]: this saves the state of a module. this is now done by *mithril*.

File: index.js [hunk 4]
```diff
@@ -134,6 +228,8 @@ function scan(render) {
   }
 
   function contains(value, el) {
+    // TODO: this is never going to be passed a vDOM and therefore will never be passed a
+    // mithril module, so how can onumloaders be added?
```
Comment (StephanHoyer) [inline index.js]: mithril itself keeps track of unloading stuff. So this might be simplified. Calling `api.onunload` might result in a route change to an empty module or so. This would result in calling `onunload` to the root-module and all components automatically.

Comment (StephanHoyer) [inline index.js]: you did exactly this. Nice

File: test.js [hunk 5]
```diff
@@ -24,7 +25,8 @@ describe('mithril query', function() {
       arrayOfArrays = m('#arrayArray');
       disabled = m('[disabled]');
       rawHtml = m.trust('<div class="trusted"></div>');
-      msxOutput = { tag: 'div', attrs: { class: 'msx' }, children: [] };
+      // TODO: I had to change class: to className: to get the msx test to pass. Is that right?
```
Comment (StephanHoyer) [inline test.js]: This was included with a PR. I don't use MSX so I did not checked this.

File: test.js [hunk 6]
```diff
@@ -321,7 +372,7 @@ describe('onunload', function() {
       return 'foo';
     }
     var out = mq(view());
-    expect(out.onunload).toNotThrow;
+    expect(out.onunload).toNotThrow();
```
Comment (StephanHoyer) [inline test.js]: did not spot this :(

**Comment by StephanHoyer** [comment]
already pretty nice.

**Comment by tombh** [comment]
OMG, I did it!

There are 3 outstanding issues for future compatibility, (see 'TODOs' in code):
1) Not all scope args (a, b, c, etc) are being passin in init().
2) `api.rootEl` no longer gets updated after a redraw, but
   `api.rootElement` does.
3) When passing mocha's `done` callback as a mithril unloader,
   mithril calls it with args which mocha doesn't like. Therefore
   `done` has to be passed as `function(){ done(); }`.

**Comment by StephanHoyer** [comment]
Sorry, I still did not found the time to test this. Hopefully I will come to this tomorrow

**Comment by tombh** [comment]
What are your thoughts on this? Still the same?

**Comment by StephanHoyer** [comment]
I'm currently in holiday with very weak internet coverage. Will look on
this in about 2 weeks.

Thomas Buckley-Houston <notifications@github.com> schrieb am Do., 6. Aug.
2015 um 12:21:

**Comment by StephanHoyer** [comment]
Just tested your implementation.

Since you also mock `window` some of my tests that also mock `window` broke. That was fixable by extending your mock.

Now I run into an issue, that a node was not found that is definetly there. I have to do futher investigation on this.

Bad thing I noticed is the performance. It's much slower.

**Comment by StephanHoyer** [comment]
I think I found the problem.

In this test I have multiple renderings at the same time:

```
      var mainOut = mq(mainView);
      var footerOut = mq(footerView);
      mainOut.setValue('input', 'foo bar');
      footerOut.click('button');
```
It seems that mq searches for `button` in `mainOut` not on `footerOut`.

Can you confirm that problem. Maybe it's good to write a test for this.

**Comment by tombh** [comment]
Sorry, just haven't had a chance to look into this. I'm still really keen on getting this merged! Hopefully I'll have a chance this week.

**Comment by tombh** [comment]
Finally found time to remind myself of all this again.

I've added a little test of what I understand of the problem, but it passes;

```
describe('querying 2 coexisting instances of mq()', function() {
    it('should query seperate instances of mq()', function() {
      var one = mq(m('input'));
      var two = mq(m('button'));
      one.should.have('input');
      two.should.have('button');
    });
  });
```

Can you provide a failing test?

**Comment by tombh** [comment]
Also, could you give me some idea of a slow test? The current suite runs in 75ms for me.

**Comment by StephanHoyer** [comment]
have you compared this to the original version?

**Comment by tombh** [comment]
I've decided to take a completely different approach I'm afraid. Instead of using mithril query I'm using jsdom.

For anyone that's interested you can see how I've done it here: https://github.com/OnApp/cloudnetv2/tree/master/frontend/spec

**Comment by StephanHoyer** [comment]
Ok, that's cool. I also tried this myself but failed... will definitely take a look.
